packages/ui-decorator/src/decorator.ts

```diff
@@ -22,19 +22,24 @@
  * SOFTWARE.
  */
 
+import type { ComponentClass } from 'react'
+
 /**
  * ---
  * category: utilities
  * ---
  * Wraps (decorates) a React component to add additional functionality
  *
  * @module decorator
- * @param {Function} decorator a function that returns a decorated component
- * @returns {Class} a decorated component class
+ * @param decorate a function that returns a decorated component
+ * @returns A decorated component class
  */
-function decorator(decorate: any) {
-  return (...args: any[]) => {
-    return (ComposedComponent: any) => {
+function decorator(
+  decorate: (toDecorate: ComponentClass<any>, ...args: any[]) => any
+) {
+  return (...args: unknown[]) => {
+    // parameters when the decorator is used
+    return (ComposedComponent: ComponentClass<any>) => {
```

Comment:
I've tried to introduce generics in many ways, so modules who use this function are giving errors when called with wrong parameters (e.g. `withStyle` is called zero parameters), but could not do it. It seems to me that TS gives up on interpreting what is going on when decorators have generics.